Packs/AgariPhishingDefense/Integrations/AgariPhishingDefense/AgariPhishingDefense.py

```diff
@@ -0,0 +1,812 @@
+        # attack_class_types Key is added to separate out the attack_class' keys in Dashboard
+        message_response['message']['attack_class_types'] = []
+        for attack_class in list(message_response['message']['attack_class'].keys()):
+            message_response['message']['attack_class_types'].append({"Types": attack_class})
+        return message_response['message']
+    except Exception as ex:
+        demisto.debug(str(ex))
+
+
+def fetch_incidents(client: Client, last_run: Dict[str, Any], args: Dict[str, Any],
+                    call_from_test=False) -> Tuple[dict, list]:
+    """
+        This function is called for fetching incidents.
+        This function gets all policies, then after using ThreadPoolExecutor, for each policy in all policies, get each
+        policy and get message from message id obtained from each policy
+        This function will execute each interval (default is 1 minute).
+
+        :param client: Client object
+        :param last_run: The greatest incident created_time we fetched from last fetch
+        :param args: The command arguments provided by user.
+        :param call_from_test: Whether calling from test module
+
+        :return next_run: This will be last_run in the next fetch-incidents
+        :return incidents: Incidents that will be created in Cortex XSOAR
+    """
+
+    # Get the last fetch time and id, if exists
+    last_fetch = last_run.get('last_fetch')
+    id = last_run.get('id', '')
+
+    first_fetch = args.get('first_fetch')
+    # Set first fetch time as default if user leave empty
+    first_fetch = DEFAULT_FIRST_FETCH if not first_fetch else first_fetch
+
+    fetch_limit = args.get('max_fetch', DEFAULT_FETCH_LIMIT)
+
+    fetch_policy_actions = args.get('fetch_policy_actions', '')
+
+    exclude_alert_type = args.get('exclude_alert_type', '')
+
+    policy_filter = args.get('policy_filter', '')
+
+    # Handle first time fetch
+    if last_fetch is None:
+        latest_created_time = dateparser.parse(first_fetch)
+    else:
+        latest_created_time = dateparser.parse(last_fetch)
+    latest_created_time = latest_created_time.strftime(DATE_FORMAT)
+
+    params = fetch_incidents_params(start_date=latest_created_time, fetch_limit=fetch_limit,
+                                    fetch_policy_actions=fetch_policy_actions, exclude_alert_type=exclude_alert_type,
+                                    policy_filter=policy_filter, id=id)
+
+    resp = get_list_policies_api_endpoint(client, params)
+
+    total_records = resp['alert_events']
+    if not total_records:
+        return last_run, []
+
+    policy_ids = []
+    policy_data = {}
+    for alert_event in total_records:
+        policy_data[alert_event['id']] = alert_event
+        policy_ids.append(alert_event['id'])
+
+    items = []
+    incidents = []
+    with ThreadPoolExecutor(max_workers=MAX_WORKERS) as executor:
```

Comment:
I suggest avoiding using parallelism within Demisto

Reply:
Agari's fetch-incident mechanism for each policy event requires 3 API calls.

1. pull policy event (10 records but 1 API Call).
2. get message link URL from each policy event. (10 API calls)
3. fetch message detail using message link's Id. (10 API calls.)

We discussed this approach with @anara123, keeping in mind that we don't have to use Demisto's global objects/variables and the number of incidents in each interval shouldn't be that long. 

We can surely discuss it on our upcoming call.